dask/array/routines.py

```diff
@@ -689,53 +716,82 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
             "See the numpy.histogram docstring for more information."
         )
 
-    if not np.iterable(bins):
-        bin_token = bins
-        mn, mx = range
-        if mn == mx:
-            mn -= 0.5
-            mx += 0.5
-
-        bins = np.linspace(mn, mx, bins + 1, endpoint=True)
-    else:
-        bin_token = bins
-    token = tokenize(a, bin_token, range, weights, density)
+    if density and isinstance(bins, Array) and bins.ndim == 0:
+        raise NotImplementedError(
+            "When `density` is True, the number of bins cannot be a delayed Dask array. "
+            "It must be a concrete number or a (possibly-delayed) array/sequence of bin edges."
+        )
 
-    nchunks = len(list(flatten(a.__dask_keys__())))
-    chunks = ((1,) * nchunks, (len(bins) - 1,))
+    if range is not None:
+        try:
+            if len(range) != 2:
```

Comment:
If range is a dataframe, this will compute (this may be unsupported/not make sense, but could still see users mistakenly doing something like this). If possible I'd like to guard `len` calls with a type check to ensure this doesn't accidentally compute something.

Reply:
I thought about something of the sort, but since this is Python, coming up with either a whitelist or blacklist of types is very hard. What if `range` is an xarray for that matter? Or a custom user type that supports the sequence interface? I feel like `isinstance(range, (list, tuple, np.ndarray, Array))` is too restrictive.

> I'd like to guard `len` calls with a type check to ensure this doesn't accidentally compute something

I agree with this though. Do you want to just explicitly blacklist `_Frame` then with a specific error message?

Reply:
That seems fair. I wouldn't worry about this then, passing in a dataframe here doesn't really make sense, so it'd be a programmer error either way. No change needed.